synapse/storage/databases/main/appservice.py

```diff
@@ -452,16 +454,15 @@ async def set_appservice_stream_type_pos(
                 % (stream_type,)
             )
 
-        def set_appservice_stream_type_pos_txn(txn: LoggingTransaction) -> None:
-            stream_id_type = "%s_stream_id" % stream_type
-            txn.execute(
-                "UPDATE application_services_state SET %s = ? WHERE as_id=?"
-                % stream_id_type,
-                (pos, service.id),
-            )
-
-        await self.db_pool.runInteraction(
-            "set_appservice_stream_type_pos", set_appservice_stream_type_pos_txn
+        # this may be the first time that we're recording any state for this AS, so
+        # we don't yet know if a row for it exists; hence we have to upsert here.
+        await self.db_pool.simple_upsert(
+            table="application_services_state",
+            keyvalues={"as_id": service.id},
+            values={f"{stream_type}_stream_id": pos},
+            # no need to lock when emulating upsert: as_id is a unique key
+            lock=False,
+            desc="set_appservice_stream_type_pos",
         )
```

Comment:
I think this is actually fixing a previously-existing race. Since the row for `service.id` wasn't added until the transaction was successfully sent - which happened asynchronously, if at all - there was no guarantee that such a row actually existed at this point.

Of course, removing the code to add the row when the txn was sent meant that the row never got added at all, so the race became a consistent failure.

TL;DR: I assert this should have been an upsert all along.